### PR TITLE
Fix Protobuf schema sampling logic

### PR DESCRIPTION
**Pull request by piochelepiotr**
# What Does This Do

When `canSample` returns false, `trySample` was not called, so the weight was not incremented.
This leads to wrong weights being calculated.
This PR updates the code to increment the weight in `canSample`, that should be called exactly once.

# Motivation

# Additional Notes

Jira ticket: [PROJ-IDENT]

### CODE REVIEW

File: dd-java-agent/instrumentation/protobuf/src/main/java/datadog/trace/instrumentation/protobuf_java/AbstractMessageInstrumentation.java [hunk 1]
```diff
@@ -36,7 +36,7 @@ public String hierarchyMarkerType() {
   public ElementMatcher<TypeDescription> hierarchyMatcher() {
     return declaresMethod(named("writeTo"))
         .and(extendsClass(named(hierarchyMarkerType())))
-        .and(not(nameStartsWith("com.google.protobuf")));
+        .and(not(nameStartsWith("com.google.protobuf")).or(named("com.google.protobuf.DynamicMessage")));
```
Comment (piochelepiotr) [inline dd-java-agent/instrumentation/protobuf/src/main/java/datadog/trace/instrumentation/protobuf_java/AbstractMessageInstrumentation.java:39]: This is another change, to track dynamic message writes. I can remove it from this PR if needed.

**Review by kr-igor — APPROVED** [review]

**Review by PerfectSlayer — APPROVED** [review]
LTGM

File: dd-java-agent/instrumentation/protobuf/src/main/java/datadog/trace/instrumentation/protobuf_java/AbstractMessageInstrumentation.java (hunk 1 above)
Comment (PerfectSlayer) [reply dd-java-agent/instrumentation/protobuf/src/main/java/datadog/trace/instrumentation/protobuf_java/AbstractMessageInstrumentation.java]: This is fine to me. You can add it to the `additional notes` section for later, if someone wonders why it was added.

File: dd-java-agent/instrumentation/protobuf/src/main/java/datadog/trace/instrumentation/protobuf_java/AbstractMessageInstrumentation.java [hunk 2]
```diff
@@ -36,7 +36,9 @@ public String hierarchyMarkerType() {
   public ElementMatcher<TypeDescription> hierarchyMatcher() {
     return declaresMethod(named("writeTo"))
         .and(extendsClass(named(hierarchyMarkerType())))
-        .and(not(nameStartsWith("com.google.protobuf")));
+        .and(
+            not(nameStartsWith("com.google.protobuf"))
+                .or(named("com.google.protobuf.DynamicMessage")));
```
Comment (vandonr) [inline dd-java-agent/instrumentation/protobuf/src/main/java/datadog/trace/instrumentation/protobuf_java/AbstractMessageInstrumentation.java:41]: wouldn't `com.google.protobuf.DynamicMessag` be included in `nameStartsWith("com.google.protobuf")` ? I feel like this second check is redundant

File: dd-trace-core/src/test/groovy/datadog/trace/core/datastreams/DefaultDataStreamsMonitoringTest.groovy [hunk 3]
```diff
@@ -77,11 +77,14 @@ class DefaultDataStreamsMonitoringTest extends DDCoreSpecification {
...
     dataStreams.trySampleSchema("schema1") == 1
+    dataStreams.canSampleSchema("schema2")
     dataStreams.trySampleSchema("schema2") == 1
-    dataStreams.trySampleSchema("schema1") == 0
-    dataStreams.trySampleSchema("schema1") == 0
+    !dataStreams.canSampleSchema("schema1")
+    !dataStreams.canSampleSchema("schema1")
     timeSource.advance(30*1e9 as long)
+    dataStreams.canSampleSchema("schema1")
     dataStreams.trySampleSchema("schema1") == 3
```
Comment (vandonr) [inline dd-trace-core/src/test/groovy/datadog/trace/core/datastreams/DefaultDataStreamsMonitoringTest.groovy:88]: I think a future maintainer would be grateful if there was one or two comments explaining why we expect those results

File: dd-java-agent/instrumentation/protobuf/src/main/java/datadog/trace/instrumentation/protobuf_java/AbstractMessageInstrumentation.java (hunk 2 above)
Comment (piochelepiotr) [reply dd-java-agent/instrumentation/protobuf/src/main/java/datadog/trace/instrumentation/protobuf_java/AbstractMessageInstrumentation.java:41]: the first one is `not`, it's to exclude a lot of the small internal classes (for example, exclude simple type serialization).

Comment (vandonr) [reply dd-java-agent/instrumentation/protobuf/src/main/java/datadog/trace/instrumentation/protobuf_java/AbstractMessageInstrumentation.java:41]: oh ok I read too fast

**Review by vandonr — APPROVED** [review]
lessgo

**Pull request merged**